cmd/memtierd/main.go

```diff
@@ -0,0 +1,339 @@
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"bufio"
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"syscall"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/containerd/nri/pkg/api"
+	"github.com/containerd/nri/pkg/stub"
+)
+
+type plugin struct {
+	stub stub.Stub
+	mask stub.EventMask
+}
+
+type MemtierdConfig struct {
+	Policy   Policy     `yaml:"policy"`
+	Routines []Routines `yaml:"routines"`
+}
+
+type Routines struct {
+	Name   string `yaml:"name"`
+	Config string `yaml:"config"`
+}
+
+type Policy struct {
+	Name   string `yaml:"name"`
+	Config string `yaml:"config"`
+}
+
+type options struct {
+	HostRoot string
+}
+
+var opt = options{}
+
+var (
+	log *logrus.Logger
+)
+
+func (p *plugin) Configure(ctx context.Context, config, runtime, version string) (stub.EventMask, error) {
+	log.Infof("Connected to %s/%s...", runtime, version)
+
+	if config == "" {
+		return 0, nil
+	}
+
+	return 0, nil
+}
+
+func (p *plugin) StartContainer(ctx context.Context, pod *api.PodSandbox, ctr *api.Container) error {
+	log.Infof("Starting container %s/%s/%s...", pod.GetNamespace(), pod.GetName(), ctr.GetName())
+
+	hostRoot := opt.HostRoot
+
+	podName := pod.GetName()
+	containerName := ctr.GetName()
+	annotations := pod.GetAnnotations()
+
+	// If memtierd annotation is not present, don't execute further
+	class, ok := annotations["class.memtierd.nri"]
+	if !ok {
+		return nil
+	}
+
+	// Check that class is of correct form
+	pattern := "^[A-Za-z0-9_-]+$"
+	regex, err := regexp.Compile(pattern)
+	if err != nil {
+		// Handle error if the pattern is invalid
+		log.Fatalf("Invalid regex pattern:", err)
+		return nil
+	}
+
+	if !regex.MatchString(class) {
+		log.Fatalf("Invalid memtierd.class.nri!")
+		return nil
+	}
+
+	// You can specify the template here based on the given class ex. class.memtierd.nri: low-prio -> template = low-prio.yaml
+	// Plugin looks in the /template directory and looks for the low-prio.yaml then
+	template := ""
+
+	if class == "example-configuration" {
+		template = "example-configuration.yaml"
+	}
+
+	fullCgroupPath := getFullCgroupPath(ctr)
+	podDirectory, outputFilePath, configFilePath := editMemtierdConfig(fullCgroupPath, podName, containerName, template, hostRoot)
+	startMemtierd(podName, containerName, podDirectory, outputFilePath, configFilePath)
+
+	return nil
+}
+
+func (p *plugin) StopContainer(ctx context.Context, pod *api.PodSandbox, ctr *api.Container) ([]*api.ContainerUpdate, error) {
+	log.Infof("Stopped container %s/%s/%s...", pod.GetNamespace(), pod.GetName(), ctr.GetName())
+
+	//
+	// This is the container (post-)stop request handler. You can update any
+	// of the remaining running containers. Take a look at the functions in
+	// pkg/api/update.go to see the available controls.
+	//
```

Comment:
I would remove this comment (which originates from the NRI template plugin), or replace it with a concise summary of what this plugin does when a container is stopped.